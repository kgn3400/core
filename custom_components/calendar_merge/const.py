"""Constants for Calendar merge integration."""

from logging import Logger, getLogger

DOMAIN = "calendar_merge"
DOMAIN_NAME = "Calendar Merge"
LOGGER: Logger = getLogger(__name__)

TRANSLATION_KEY = DOMAIN
TRANSLATION_KEY_MISSING_ENTITY = "missing_entity"
TRANSLATION_KEY_MISSING__TIMER_ENTITY = "missing_timer_entity"
TRANSLATION_KEY_TEMPLATE_ERROR = "template_error"

CONF_DAYS_AHEAD = "days_ahead"
CONF_MAX_EVENTS = "max_events"
CONF_CALENDAR_ENTITY_IDS = "calender_entity_ids"
CONF_REMOVE_RECURRING_EVENTS = "remove_recurring_events"
CONF_SHOW_EVENT_AS_TIME_TO = "show_event_as_time_to"
CONF_SHOW_END_DATE = "show_show_end_date"
CONF_SHOW_SUMMARY = "show_summary"
CONF_USE_SUMMARY_AS_ENTITY_NAME = "use_summary_as_entity_name"
CONF_FORMAT_LANGUAGE = "format_language"

CONF_MD_HEADER_TEMPLATE = "md_header_template"
CONF_DEFAULT_MD_HEADER_TEMPLATE = "defaults.default_md_header_template"

CONF_MD_ITEM_TEMPLATE = "md_item_template"
CONF_DEFAULT_MD_ITEM_TEMPLATE = "defaults.default_md_item_template"

SERVICE_SAVE_SETTINGS = "save_settings"
