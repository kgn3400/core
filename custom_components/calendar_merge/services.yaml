# Service ID
toggle_show_as_time_to:
  target:
    entity:
      integration: calendar_merge
      domain: sensor
      supported_features:
        - fan.FanEntityFeature.SET_SPEED
  fields:
    # Key of the field
    save_settings:
      # Field name as shown in UI
      # name: Order by
      # Description of the field
      # description: Order message list by
      # Whether or not field is required (default = false)
      required: false
      # Advanced fields are only shown when the advanced mode is enabled for the user
      # (default = false)
      example: false
      # The default field value
      default: false
      selector:
        boolean:
