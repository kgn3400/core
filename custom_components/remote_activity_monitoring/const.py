"""Constants for Ping alert integration."""

from enum import StrEnum
from logging import Logger, getLogger

DOMAIN = "remote_activity_monitoring"
DOMAIN_NAME = "Remote activity monitoring"
LOGGER: Logger = getLogger(__name__)

TRANSLATION_KEY = DOMAIN
TRANSLATION_KEY_MISSING_ENTITY = "missing_entity"
TRANSLATION_KEY_MISSING__TIMER_ENTITY = "missing_timer_entity"
TRANSLATION_KEY_TEMPLATE_ERROR = "template_error"

CONF_COMPONENT_TYPE = "component_type"
CONF_SECURE = "secure"
CONF_MONITOR_ENTITY = "monitor_entity"


CONF_ENTITY_IDS = "entity_ids"


class ComponentType(StrEnum):
    """Available entity component types."""

    MAIN = "main"
    REMOTE = "remote"


class StepType(StrEnum):
    """Available entity component types."""

    CONFIG = "config"
    OPTIONS = "options"
