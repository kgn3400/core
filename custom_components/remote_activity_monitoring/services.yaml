# Service ID
get_remotes:
# Service name as shown in UI
# name: Update
# Description of the service
# description: Check for PyPi updates.

