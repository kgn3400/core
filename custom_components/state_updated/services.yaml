# Service ID
reset_all:
  # Service name as shown in UI
#  name: Reset all
# Description of the service
#  description: Reset state updated for all entities.
reset_entity:
  target:
    entity:
      integration: state_updated
      # domain: state_updated
  # fields:
  #   command:
  #     required: true
  #     example: "HOME"
  #     selector:
  #       text:
