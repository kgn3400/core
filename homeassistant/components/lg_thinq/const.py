"""Constants for LG ThinQ."""

# Base component constants.
from typing import Final

from thinqconnect import (
    AirConditionerDevice,
    AirPurifierDevice,
    AirPurifierFanDevice,
    CeilingFanDevice,
    CooktopDevice,
    DehumidifierDevice,
    DeviceType,
    DishWasherDevice,
    DryerDevice,
    HomeBrewDevice,
    HoodDevice,
    HumidifierDevice,
    KimchiRefrigeratorDevice,
    MicrowaveOvenDevice,
    OvenDevice,
    PlantCultivatorDevice,
    RefrigeratorDevice,
    RobotCleanerDevice,
    StickCleanerDevice,
    StylerDevice,
    SystemBoilerDevice,
    WashcomboMainDevice,
    WashcomboMiniDevice,
    WasherDevice,
    WashtowerDevice,
    WashtowerDryerDevice,
    WashtowerWasherDevice,
    WaterHeaterDevice,
    WaterPurifierDevice,
    WineCellarDevice,
)

# Common
DOMAIN = "lg_thinq"
COMPANY = "LGE"
THINQ_DEFAULT_NAME: Final = "LG ThinQ"
THINQ_PAT_URL: Final = "https://connect-pat.lgthinq.com"

# Config Flow
CLIENT_PREFIX: Final = "home-assistant"
CONF_CONNECT_CLIENT_ID: Final = "connect_client_id"
DEFAULT_COUNTRY: Final = "US"

THINQ_DEVICE_ADDED: Final = "thinq_device_added"

DEVICE_TYPE_API_MAP: Final = {
    DeviceType.AIR_CONDITIONER: AirConditionerDevice,
    DeviceType.AIR_PURIFIER_FAN: AirPurifierFanDevice,
    DeviceType.AIR_PURIFIER: AirPurifierDevice,
    DeviceType.CEILING_FAN: CeilingFanDevice,
    DeviceType.COOKTOP: CooktopDevice,
    DeviceType.DEHUMIDIFIER: DehumidifierDevice,
    DeviceType.DISH_WASHER: DishWasherDevice,
    DeviceType.DRYER: DryerDevice,
    DeviceType.HOME_BREW: HomeBrewDevice,
    DeviceType.HOOD: HoodDevice,
    DeviceType.HUMIDIFIER: HumidifierDevice,
    DeviceType.KIMCHI_REFRIGERATOR: KimchiRefrigeratorDevice,
    DeviceType.MICROWAVE_OVEN: MicrowaveOvenDevice,
    DeviceType.OVEN: OvenDevice,
    DeviceType.PLANT_CULTIVATOR: PlantCultivatorDevice,
    DeviceType.REFRIGERATOR: RefrigeratorDevice,
    DeviceType.ROBOT_CLEANER: RobotCleanerDevice,
    DeviceType.STICK_CLEANER: StickCleanerDevice,
    DeviceType.STYLER: StylerDevice,
    DeviceType.SYSTEM_BOILER: SystemBoilerDevice,
    DeviceType.WASHER: WasherDevice,
    DeviceType.WASHCOMBO_MAIN: WashcomboMainDevice,
    DeviceType.WASHCOMBO_MINI: WashcomboMiniDevice,
    DeviceType.WASHTOWER_DRYER: WashtowerDryerDevice,
    DeviceType.WASHTOWER: WashtowerDevice,
    DeviceType.WASHTOWER_WASHER: WashtowerWasherDevice,
    DeviceType.WATER_HEATER: WaterHeaterDevice,
    DeviceType.WATER_PURIFIER: WaterPurifierDevice,
    DeviceType.WINE_CELLAR: WineCellarDevice,
}
