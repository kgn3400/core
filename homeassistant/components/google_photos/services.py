"""Google Photos services."""

from __future__ import annotations

import asyncio
import mimetypes
from pathlib import Path

import voluptuous as vol

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_FILENAME
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
)
from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
from homeassistant.helpers import config_validation as cv

from . import api
from .const import DOMAIN, UPLOAD_SCOPE

type GooglePhotosConfigEntry = ConfigEntry[api.AsyncConfigEntryAuth]

__all__ = [
    "DOMAIN",
]

CONF_CONFIG_ENTRY_ID = "config_entry_id"

UPLOAD_SERVICE = "upload"
UPLOAD_SERVICE_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_CONFIG_ENTRY_ID): cv.string,
        vol.Required(CONF_FILENAME): vol.All(cv.ensure_list, [cv.string]),
    }
)


def _read_file_contents(
    hass: HomeAssistant, filenames: list[str]
) -> list[tuple[str, bytes]]:
    """Return the mime types and file contents for each file."""
    results = []
    for filename in filenames:
        if not hass.config.is_allowed_path(filename):
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="no_access_to_path",
                translation_placeholders={"filename": filename},
            )
        filename_path = Path(filename)
        if not filename_path.exists():
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="filename_does_not_exist",
                translation_placeholders={"filename": filename},
            )
        mime_type, _ = mimetypes.guess_type(filename)
        if mime_type is None or not (mime_type.startswith(("image", "video"))):
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="filename_is_not_image",
                translation_placeholders={"filename": filename},
            )
        results.append((mime_type, filename_path.read_bytes()))
    return results


def async_register_services(hass: HomeAssistant) -> None:
    """Register Google Photos services."""

    async def async_handle_upload(call: ServiceCall) -> ServiceResponse:
        """Generate content from text and optionally images."""
        config_entry: GooglePhotosConfigEntry | None = (
            hass.config_entries.async_get_entry(call.data[CONF_CONFIG_ENTRY_ID])
        )
        if not config_entry:
            raise ServiceValidationError(
                translation_domain=DOMAIN,
                translation_key="integration_not_found",
                translation_placeholders={"target": DOMAIN},
            )
        scopes = config_entry.data["token"]["scope"].split(" ")
        if UPLOAD_SCOPE not in scopes:
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="missing_upload_permission",
                translation_placeholders={"target": DOMAIN},
            )

        client_api = config_entry.runtime_data
        upload_tasks = []
        file_results = await hass.async_add_executor_job(
            _read_file_contents, hass, call.data[CONF_FILENAME]
        )
        for mime_type, content in file_results:
            upload_tasks.append(client_api.upload_content(content, mime_type))
        upload_tokens = await asyncio.gather(*upload_tasks)
        media_ids = await client_api.create_media_items(upload_tokens)
        if call.return_response:
            return {
                "media_items": [{"media_item_id": media_id for media_id in media_ids}]
            }
        return None

    if not hass.services.has_service(DOMAIN, UPLOAD_SERVICE):
        hass.services.async_register(
            DOMAIN,
            UPLOAD_SERVICE,
            async_handle_upload,
            schema=UPLOAD_SERVICE_SCHEMA,
            supports_response=SupportsResponse.OPTIONAL,
        )
