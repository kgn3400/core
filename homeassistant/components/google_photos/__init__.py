"""The Google Photos integration."""

from __future__ import annotations

from aiohttp import ClientError, ClientResponseError

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers import config_entry_oauth2_flow

from . import api
from .const import DOMAIN
from .services import async_register_services

type GooglePhotosConfigEntry = ConfigEntry[api.AsyncConfigEntryAuth]

__all__ = [
    "DOMAIN",
]


async def async_setup_entry(
    hass: HomeAssistant, entry: GooglePhotosConfigEntry
) -> bool:
    """Set up Google Photos from a config entry."""
    implementation = (
        await config_entry_oauth2_flow.async_get_config_entry_implementation(
            hass, entry
        )
    )
    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
    auth = api.AsyncConfigEntryAuth(hass, session)
    try:
        await auth.async_get_access_token()
    except ClientResponseError as err:
        if 400 <= err.status < 500:
            raise ConfigEntryAuthFailed(
                "OAuth session is not valid, reauth required"
            ) from err
        raise ConfigEntryNotReady from err
    except ClientError as err:
        raise ConfigEntryNotReady from err
    entry.runtime_data = auth

    async_register_services(hass)

    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: GooglePhotosConfigEntry
) -> bool:
    """Unload a config entry."""
    return True
