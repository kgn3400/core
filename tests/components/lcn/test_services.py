"""Test for the LCN services."""

from unittest.mock import patch

import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.lcn import DOMAIN
from homeassistant.components.lcn.const import (
    CONF_KEYS,
    CONF_LED,
    CONF_OUTPUT,
    CONF_PCK,
    CONF_RELVARREF,
    CONF_ROW,
    CONF_SETPOINT,
    CONF_TABLE,
    CONF_TEXT,
    CONF_TIME,
    CONF_TIME_UNIT,
    CONF_TRANSITION,
    CONF_VALUE,
    CONF_VARIABLE,
)
from homeassistant.components.lcn.services import LcnService
from homeassistant.const import (
    CONF_ADDRESS,
    CONF_BRIGHTNESS,
    CONF_STATE,
    CONF_UNIT_OF_MEASUREMENT,
)
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component

from .conftest import MockModuleConnection, MockPchkConnectionManager, setup_component


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_abs(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_abs service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "dim_output") as dim_output:
        await hass.services.async_call(
            DOMAIN,
            LcnService.OUTPUT_ABS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_OUTPUT: "output1",
                CONF_BRIGHTNESS: 100,
                CONF_TRANSITION: 5,
            },
            blocking=True,
        )

    assert dim_output.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_rel(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_rel service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "rel_output") as rel_output:
        await hass.services.async_call(
            DOMAIN,
            LcnService.OUTPUT_REL,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_OUTPUT: "output1",
                CONF_BRIGHTNESS: 25,
            },
            blocking=True,
        )

    assert rel_output.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_output_toggle(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test output_toggle service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "toggle_output") as toggle_output:
        await hass.services.async_call(
            DOMAIN,
            LcnService.OUTPUT_TOGGLE,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_OUTPUT: "output1",
                CONF_TRANSITION: 5,
            },
            blocking=True,
        )

    assert toggle_output.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_relays(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test relays service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "control_relays") as control_relays:
        await hass.services.async_call(
            DOMAIN,
            LcnService.RELAYS,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_STATE: "0011TT--"},
            blocking=True,
        )

    assert control_relays.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_led(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test led service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "control_led") as control_led:
        await hass.services.async_call(
            DOMAIN,
            LcnService.LED,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_LED: "led6", CONF_STATE: "blink"},
            blocking=True,
        )

    assert control_led.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_abs(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_abs service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_abs") as var_abs:
        await hass.services.async_call(
            DOMAIN,
            LcnService.VAR_ABS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_VARIABLE: "var1",
                CONF_VALUE: 75,
                CONF_UNIT_OF_MEASUREMENT: "%",
            },
            blocking=True,
        )

    assert var_abs.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_rel(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_rel service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_rel") as var_rel:
        await hass.services.async_call(
            DOMAIN,
            LcnService.VAR_REL,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_VARIABLE: "var1",
                CONF_VALUE: 10,
                CONF_UNIT_OF_MEASUREMENT: "%",
                CONF_RELVARREF: "current",
            },
            blocking=True,
        )

    assert var_rel.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_var_reset(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test var_reset service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "var_reset") as var_reset:
        await hass.services.async_call(
            DOMAIN,
            LcnService.VAR_RESET,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_VARIABLE: "var1"},
            blocking=True,
        )

    assert var_reset.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_regulator(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_regulator service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "lock_regulator") as lock_regulator:
        await hass.services.async_call(
            DOMAIN,
            LcnService.LOCK_REGULATOR,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_SETPOINT: "r1varsetpoint",
                CONF_STATE: True,
            },
            blocking=True,
        )

    assert lock_regulator.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_send_keys(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test send_keys service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "send_keys") as send_keys:
        await hass.services.async_call(
            DOMAIN,
            LcnService.SEND_KEYS,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_KEYS: "a1a5d8", CONF_STATE: "hit"},
            blocking=True,
        )

    keys = [[False] * 8 for i in range(4)]
    keys[0][0] = True
    keys[0][4] = True
    keys[3][7] = True

    assert send_keys.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_send_keys_hit_deferred(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test send_keys (hit_deferred) service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    keys = [[False] * 8 for i in range(4)]
    keys[0][0] = True
    keys[0][4] = True
    keys[3][7] = True

    # success
    with patch.object(
        MockModuleConnection, "send_keys_hit_deferred"
    ) as send_keys_hit_deferred:
        await hass.services.async_call(
            DOMAIN,
            LcnService.SEND_KEYS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_KEYS: "a1a5d8",
                CONF_TIME: 5,
                CONF_TIME_UNIT: "s",
            },
            blocking=True,
        )

    assert send_keys_hit_deferred.await_args.args == snapshot()

    # wrong key action
    with (
        patch.object(
            MockModuleConnection, "send_keys_hit_deferred"
        ) as send_keys_hit_deferred,
        pytest.raises(ValueError),
    ):
        await hass.services.async_call(
            DOMAIN,
            LcnService.SEND_KEYS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_KEYS: "a1a5d8",
                CONF_STATE: "make",
                CONF_TIME: 5,
                CONF_TIME_UNIT: "s",
            },
            blocking=True,
        )


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_keys(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_keys service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "lock_keys") as lock_keys:
        await hass.services.async_call(
            DOMAIN,
            LcnService.LOCK_KEYS,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_TABLE: "a", CONF_STATE: "0011TT--"},
            blocking=True,
        )

    assert lock_keys.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_lock_keys_tab_a_temporary(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test lock_keys (tab_a_temporary) service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    # success
    with patch.object(
        MockModuleConnection, "lock_keys_tab_a_temporary"
    ) as lock_keys_tab_a_temporary:
        await hass.services.async_call(
            DOMAIN,
            LcnService.LOCK_KEYS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_STATE: "0011TT--",
                CONF_TIME: 10,
                CONF_TIME_UNIT: "s",
            },
            blocking=True,
        )

    assert lock_keys_tab_a_temporary.await_args.args == snapshot()

    # wrong table
    with (
        patch.object(
            MockModuleConnection, "lock_keys_tab_a_temporary"
        ) as lock_keys_tab_a_temporary,
        pytest.raises(ValueError),
    ):
        await hass.services.async_call(
            DOMAIN,
            LcnService.LOCK_KEYS,
            {
                CONF_ADDRESS: "pchk.s0.m7",
                CONF_TABLE: "b",
                CONF_STATE: "0011TT--",
                CONF_TIME: 10,
                CONF_TIME_UNIT: "s",
            },
            blocking=True,
        )


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_dyn_text(
    hass: HomeAssistant, snapshot: SnapshotAssertion
) -> None:
    """Test dyn_text service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "dyn_text") as dyn_text:
        await hass.services.async_call(
            DOMAIN,
            LcnService.DYN_TEXT,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_ROW: 1, CONF_TEXT: "text in row 1"},
            blocking=True,
        )

    assert dyn_text.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_pck(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test pck service."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "pck") as pck:
        await hass.services.async_call(
            DOMAIN,
            LcnService.PCK,
            {CONF_ADDRESS: "pchk.s0.m7", CONF_PCK: "PIN4"},
            blocking=True,
        )

    assert pck.await_args.args == snapshot()


@patch("pypck.connection.PchkConnectionManager", MockPchkConnectionManager)
async def test_service_called_with_invalid_host_id(hass: HomeAssistant) -> None:
    """Test service was called with non existing host id."""
    await async_setup_component(hass, "persistent_notification", {})
    await setup_component(hass)

    with patch.object(MockModuleConnection, "pck") as pck, pytest.raises(ValueError):
        await hass.services.async_call(
            DOMAIN,
            LcnService.PCK,
            {CONF_ADDRESS: "foobar.s0.m7", CONF_PCK: "PIN4"},
            blocking=True,
        )

    pck.assert_not_awaited()
