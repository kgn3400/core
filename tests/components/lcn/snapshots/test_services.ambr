# serializer version: 1
# name: test_service_dyn_text
  tuple(
    0,
    'text in row 1',
  )
# ---
# name: test_service_led
  tuple(
    <LedPort.LED6: 5>,
    <LedStatus.BLINK: 'B'>,
  )
# ---
# name: test_service_lock_keys
  tuple(
    0,
    list([
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_lock_keys_tab_a_temporary
  tuple(
    10,
    <TimeUnit.SECONDS: 'S'>,
    list([
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.OFF: '0'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.ON: '1'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.TOGGLE: 'U'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
      <KeyLockStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_lock_regulator
  tuple(
    0,
    True,
  )
# ---
# name: test_service_output_abs
  tuple(
    0,
    100,
    9,
  )
# ---
# name: test_service_output_rel
  tuple(
    0,
    25,
  )
# ---
# name: test_service_output_toggle
  tuple(
    0,
    9,
  )
# ---
# name: test_service_pck
  tuple(
    'PIN4',
  )
# ---
# name: test_service_relays
  tuple(
    list([
      <RelayStateModifier.OFF: '0'>,
      <RelayStateModifier.OFF: '0'>,
      <RelayStateModifier.ON: '1'>,
      <RelayStateModifier.ON: '1'>,
      <RelayStateModifier.TOGGLE: 'U'>,
      <RelayStateModifier.TOGGLE: 'U'>,
      <RelayStateModifier.NOCHANGE: '-'>,
      <RelayStateModifier.NOCHANGE: '-'>,
    ]),
  )
# ---
# name: test_service_send_keys
  tuple(
    list([
      list([
        True,
        False,
        False,
        False,
        True,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        True,
      ]),
    ]),
    <SendKeyCommand.HIT: 'K'>,
  )
# ---
# name: test_service_send_keys_hit_deferred
  tuple(
    list([
      list([
        True,
        False,
        False,
        False,
        True,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        False,
      ]),
      list([
        False,
        False,
        False,
        False,
        False,
        False,
        False,
        True,
      ]),
    ]),
    5,
    <TimeUnit.SECONDS: 'S'>,
  )
# ---
# name: test_service_var_abs
  tuple(
    <Var.VAR1ORTVAR: 0>,
    75.0,
    <VarUnit.PERCENT: '%'>,
  )
# ---
# name: test_service_var_rel
  tuple(
    <Var.VAR1ORTVAR: 0>,
    10.0,
    <VarUnit.PERCENT: '%'>,
    <RelVarRef.CURRENT: 1>,
  )
# ---
# name: test_service_var_reset
  tuple(
    <Var.VAR1ORTVAR: 0>,
  )
# ---
